exporter/cassandra/exporter.go

```diff
@@ -0,0 +1,164 @@
+package cassandra
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/jaeger"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger-opentelemetry-collector/internal/metrics"
+	"github.com/jaegertracing/jaeger/pkg/cassandra"
+	casConfig "github.com/jaegertracing/jaeger/pkg/cassandra/config"
+	plugin "github.com/jaegertracing/jaeger/plugin/storage/cassandra"
+	cSpanStore "github.com/jaegertracing/jaeger/plugin/storage/cassandra/spanstore"
+	"github.com/jaegertracing/jaeger/plugin/storage/cassandra/spanstore/dbmodel"
+)
+
+func newTracesExporter(cfg *Config, set component.ExporterCreateSettings) (component.TracesExporter, error) {
+	s := newSender(cfg, set.TelemetrySettings)
+	return newTracesExporterWithSender(cfg, set, s)
+}
+
+func newTracesExporterWithSender(cfg *Config, set component.ExporterCreateSettings, sender *sender) (component.TracesExporter, error) {
+	return exporterhelper.NewTracesExporter(
+		context.TODO(), set, cfg, sender.pushTraces,
+		exporterhelper.WithCapabilities(consumer.Capabilities{MutatesData: false}),
+		exporterhelper.WithStart(sender.start),
+		exporterhelper.WithShutdown(sender.shutdown),
+		exporterhelper.WithTimeout(cfg.TimeoutSettings),
+		exporterhelper.WithRetry(cfg.RetrySettings),
+		exporterhelper.WithQueue(cfg.QueueSettings),
+	)
+}
+
+type sender struct {
+	name           string
+	settings       component.TelemetrySettings
+	cfg            *Config
+	primaryConfig  casConfig.SessionBuilder
+	primarySession cassandra.Session
+	// factory  *plugin.Factory // Cannot use unless some functions or fields are exported
+	writer *cSpanStore.SpanWriter
+}
+
+func newSender(cfg *Config, settings component.TelemetrySettings) *sender {
+	return &sender{
+		name:          cfg.ID().String(),
+		settings:      settings,
+		cfg:           cfg,
+		primaryConfig: cfg.Options.GetPrimary(),
+	}
+}
+
+func (s *sender) pushTraces(ctx context.Context, td ptrace.Traces) error {
+	batches, err := jaeger.ProtoFromTraces(td)
+	if err != nil {
+		return consumererror.NewPermanent(fmt.Errorf("failed to push trace data via cassandra exporter: %w", err))
+	}
+
+	for _, batch := range batches {
+		process := batch.GetProcess()
+		if process == nil {
+			s.settings.Logger.Debug("NO PROCESS")
+		}
+		for _, span := range batch.GetSpans() {
+			span.Process = process
+			err = s.writer.WriteSpan(ctx, span)
```

Comment:
So the approach uses the existing Jaeger Cassandra writer. 

More preferable approach would be to have a native implementation and convert OTLP directly to the Cassandra model.

Reply:
I agree, there's a few rounds of conversion that we could do just once. This is mostly me hooking up two hoses that fit together. There is a considerable amount of work I believe to make it work the way we believe it should. I would prefer to use other people's (smarter people's) code and having done this I wonder if the better way to go about this is to actually submit a new exporter to the contrib repository. That gives us access to internal libaries for translation and testing that we don't have access to here.

For instance much of the code in the translator would be great to have https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/translator/jaeger/traces_to_jaegerproto.go 

Thoughts? Otherwise I'll do what I'm told and try it here.

Reply:
The translator from contrib will not help us with the native implementation that directly translates OTLP to the Jaeger cassandra model https://github.com/jaegertracing/jaeger/blob/8611137adaffb9c661845109462408d5499b47dc/plugin/storage/cassandra/spanstore/dbmodel/model.go#L40. 

In the past we have as well talked about potentially defining new Cassanda model that is closer to OTLP if there are valid arguments to do so.